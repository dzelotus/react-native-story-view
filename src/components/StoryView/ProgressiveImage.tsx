import React from 'react';
import { Animated, View } from 'react-native';
import styles from './styles';
import type { ProgressiveImageProps } from './types';

const ProgressiveImage = (props: ProgressiveImageProps) => {
  const thumbnailAnimated = new Animated.Value(0.2);
  const { thumbnailSource, imgSource, ...reset } = props;

  const imageAnimated = new Animated.Value(1);

  const handleThumbnailLoad = () => {
    Animated.timing(thumbnailAnimated, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  const onImageLoad = () => {
    Animated.timing(imageAnimated, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  return (
    <View style={styles.progressiveImageContainer}>
      <Animated.Image
        {...reset}
        source={thumbnailSource}
        style={[
          styles.imageOverlay,
          { opacity: thumbnailAnimated, height: '100%' },
        ]}
        onLoad={handleThumbnailLoad}
        resizeMethod="resize"
        resizeMode="cover"
      />
      <Animated.Image
        {...reset}
        source={imgSource}
        style={[{ opacity: imageAnimated, height: '100%' }]}
        onLoad={onImageLoad}
        onLoadEnd={() => props.onImageLoaded && props.onImageLoaded()}
        resizeMethod="resize"
        resizeMode="cover"
      />
    </View>
  );
};

export default ProgressiveImage;
