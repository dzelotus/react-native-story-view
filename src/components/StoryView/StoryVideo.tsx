import React, { useEffect, useRef, useState } from 'react';
import {
  View,
  ActivityIndicator,
  Platform,
  ActivityIndicatorProps,
} from 'react-native';
import Video, { VideoRef, OnBufferData } from 'react-native-video';
import type { OnLoadData, OnProgressData } from 'react-native-video';
import styles from './styles';
import { Colors } from '../../theme';
import type { StoryType } from './types';

const MIN_BUFFER_TIME = 1000 * 1;
const PlAYBACK_BUFFER_TIME = 1000 * 1;
const PLAYBACK_AFTER_REBUFFER = 1000 * 1;

interface StoryVideoProps {
  source: StoryType;
  pause: boolean;
  index?: number;
  storyIndex?: number;
  setPause?: (value: boolean) => void;
  onVideoEnd?: () => void;
  onVideoProgress?: (progressData?: OnProgressData) => void;
  onVideoLoaded?: (arg0: OnLoadData) => void;
  videoProps: any;
  showSourceIndicator?: boolean;
  sourceIndicatorProps?: ActivityIndicatorProps;
}

const StoryVideo = ({
  source,
  pause,
  index,
  storyIndex,
  setPause,
  onVideoEnd,
  onVideoProgress,
  onVideoLoaded,
  sourceIndicatorProps,
}: StoryVideoProps) => {
  const [loading, setLoading] = useState(true);
  const [buffering, setBuffering] = useState(true);

  const videoRef = useRef<VideoRef>(null);
  const videoData = useRef<OnLoadData>();

  const isCurrentIndex = index === storyIndex;

  useEffect(() => {
    if (index === storyIndex) {
      if (Platform.OS === 'android' || source?.url?.endsWith('m3u8')) {
        videoRef?.current?.seek(0);
      } else {
        videoRef?.current?.seek(1);
      }
    }
  }, [storyIndex, index, source?.url]);

  const loadVideo = async () => {
    if (isCurrentIndex) {
      if (videoData.current === undefined) {
        return;
      }
      if (loading) {
        setLoading(false);
      }
      if (buffering) {
        setBuffering(false);
      }
    }
  };

  const onBuffer = (data: OnBufferData) => {
    setBuffering(data.isBuffering);
  };

  return (
    <View style={{ flex: 1 }}>
      <Video
        ref={videoRef}
        resizeMode="cover"
        paused={pause || loading}
        source={{ uri: source?.url }}
        onSeek={event => {
          if (index === storyIndex) {
            if (Platform.OS === 'ios' && !source?.url?.endsWith('m3u8')) {
              if (event.seekTime === 1) {
                videoRef?.current?.seek(0.15);
                setPause?.(true);
                setPause?.(false);
                videoRef?.current?.resume();
              } else {
                videoRef?.current?.resume();
              }
            }
          }
        }}
        onEnd={onVideoEnd}
        onError={(_error: any) => {
          setLoading(false);
        }}
        onProgress={data => {
          if (isCurrentIndex) {
            onVideoProgress?.(data);
          }
        }}
        bufferConfig={{
          minBufferMs: MIN_BUFFER_TIME,
          bufferForPlaybackMs: PlAYBACK_BUFFER_TIME,
          bufferForPlaybackAfterRebufferMs: PLAYBACK_AFTER_REBUFFER,
        }}
        onBuffer={onBuffer}
        onLoad={(item: OnLoadData) => {
          videoData.current = item;
          videoRef?.current?.resume();
          loadVideo();
          onVideoLoaded?.(item);
        }}
        style={styles.contentVideoView}
      />
      {(loading || buffering) && (
        <ActivityIndicator
          animating
          pointerEvents="none"
          color={Colors.loaderColor}
          size="small"
          style={styles.loaderView}
          {...sourceIndicatorProps}
        />
      )}
    </View>
  );
};

export default React.memo(StoryVideo);
