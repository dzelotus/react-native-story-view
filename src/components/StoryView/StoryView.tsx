import React from 'react';
import { useWindowDimensions, View } from 'react-native';
import ProgressiveImage from './ProgressiveImage';
import styles from './styles';
import { StoryViewProps, StroyTypes } from './types';
import StoryVideo from './StoryVideo';

const StoryView = (props: StoryViewProps) => {
  const source = props?.stories?.[props?.progressIndex];
  const isCurrentIndex = props?.index === props?.storyIndex;

  const { height, width } = useWindowDimensions();

  return (
    <View style={[styles.divStory, { height, width }]} ref={props?.viewRef}>
      {source?.type === StroyTypes.Image && isCurrentIndex ? (
        <ProgressiveImage
          viewStyle={props?.imageStyle ?? styles.imgStyle}
          imgSource={{ uri: source.url ?? '' }}
          thumbnailSource={{ uri: source.url ?? '' }}
          onImageLoaded={props.onImageLoaded}
        />
      ) : (
        isCurrentIndex && (
          <StoryVideo
            source={source}
            pause={!!props.pause}
            index={props?.index}
            storyIndex={props?.storyIndex}
            setPause={props?.setPause}
            onVideoEnd={props?.onVideoEnd}
            onVideoProgress={props?.onVideoProgress}
            showSourceIndicator={props?.showSourceIndicator}
            sourceIndicatorProps={props?.sourceIndicatorProps}
            videoProps={props?.videoProps}
          />
        )
      )}
    </View>
  );
};

export default StoryView;
