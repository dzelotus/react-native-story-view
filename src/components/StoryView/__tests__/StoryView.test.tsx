import React from 'react';
import { render } from '@testing-library/react-native';
import StoryView from '../StoryView';
import StoryContainer from '../StoryContainer';
import ProgressView from '../ProgressView';
import { Footer } from '../../Footer';

jest.useFakeTimers();
jest.mock('react-native-video', () => 'Video');

describe('StoryView component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <StoryView videoDuration={[]} stories={[]} progressIndex={0} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('StoryContainer component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <StoryContainer visible stories={[]} maxVideoDuration={5} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('ProgressView component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <ProgressView
        storyIndex={0}
        isLoaded
        duration={10}
        pause={false}
        stories={[]}
        currentIndex={0}
        currentStory={{}}
        length={[]}
        progress={{}}
        videoDuration={[]}
        setVideoDuration={() => {}}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('StoryView should render Footer component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <StoryContainer stories={[]} renderFooterComponent={() => <Footer />} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
