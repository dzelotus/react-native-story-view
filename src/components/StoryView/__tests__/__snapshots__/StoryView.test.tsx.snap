// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressView component Match Snapshot 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 16.428571428571427,
      "justifyContent": "space-between",
      "left": 10,
      "position": "absolute",
      "right": 10,
      "top": 16.428571428571427,
    }
  }
/>
`;

exports[`StoryContainer component Match Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#00000000",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          onLayout={[Function]}
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignSelf": "center",
                    "height": "100%",
                    "paddingBottom": 6,
                    "width": 750,
                  },
                  Object {
                    "height": 1334,
                    "width": 750,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Video
                  bufferConfig={
                    Object {
                      "bufferForPlaybackAfterRebufferMs": 1000,
                      "bufferForPlaybackMs": 1000,
                      "minBufferMs": 1000,
                    }
                  }
                  onBuffer={[Function]}
                  onEnd={[Function]}
                  onError={[Function]}
                  onLoad={[Function]}
                  onProgress={[Function]}
                  onSeek={[Function]}
                  paused={true}
                  resizeMode="cover"
                  source={
                    Object {
                      "uri": undefined,
                    }
                  }
                  style={
                    Object {
                      "backgroundColor": "black",
                      "borderBottomLeftRadius": 16,
                      "borderBottomRightRadius": 16,
                      "flex": 1,
                      "overflow": "hidden",
                    }
                  }
                />
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "bottom": 0,
                      "justifyContent": "center",
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "red",
                        "fontSize": 16,
                      }
                    }
                  />
                </View>
                <ActivityIndicator
                  animating={true}
                  color="#575899"
                  pointerEvents="none"
                  size="small"
                  style={
                    Object {
                      "flex": 1,
                      "left": "45%",
                      "position": "absolute",
                      "top": "50%",
                    }
                  }
                />
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "position": "absolute",
                  "width": "100%",
                },
                Object {
                  "opacity": 1,
                },
              ]
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 16.428571428571427,
                  "justifyContent": "space-between",
                  "left": 10,
                  "position": "absolute",
                  "right": 10,
                  "top": 16.428571428571427,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`StoryView component Match Snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "center",
        "height": "100%",
        "paddingBottom": 6,
        "width": 750,
      },
      Object {
        "height": 1334,
        "width": 750,
      },
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Video
      bufferConfig={
        Object {
          "bufferForPlaybackAfterRebufferMs": 1000,
          "bufferForPlaybackMs": 1000,
          "minBufferMs": 1000,
        }
      }
      onBuffer={[Function]}
      onError={[Function]}
      onLoad={[Function]}
      onProgress={[Function]}
      onSeek={[Function]}
      paused={true}
      resizeMode="cover"
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "backgroundColor": "black",
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "flex": 1,
          "overflow": "hidden",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "red",
            "fontSize": 16,
          }
        }
      />
    </View>
    <ActivityIndicator
      animating={true}
      color="#575899"
      pointerEvents="none"
      size="small"
      style={
        Object {
          "flex": 1,
          "left": "45%",
          "position": "absolute",
          "top": "50%",
        }
      }
    />
  </View>
</View>
`;

exports[`StoryView should render Footer component Match Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  />
</RCTSafeAreaView>
`;
